src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -447,3 +428,21 @@ pub fn cleanup_kinds(mir: &mir::Body<'_>) -> IndexVec<mir::BasicBlock, CleanupKi
     debug!("cleanup_kinds: result={:?}", result);
     result
 }
+
+/// Returns `true` if locals of this type need to be allocated on the stack.
+fn ty_requires_alloca<'a, 'tcx>(
+    fx: &FunctionCx<'a, 'tcx, impl BuilderMethods<'a, 'tcx>>,
+    ty: TyAndLayout<'tcx>,
+) -> bool {
+    // Currently, this returns `true` for ADTs that are otherwise small enough to qualify. For
+    // example, `struct Newtype(i32)`. This way, every type has a single way to extract data
+    // (gep, extractvalue, etc.).
```

Comment:
This comment appears to be rewritten, but hasn't been true in ages, newtypes of immediates (scalars/vectors) and scalar pairs get "unpacked" automatically.